Support graphql-ws protocol, breaks existing subscriptions-transport-ws protocol

ropwareJB opened the pull request:
+ Bump support to newer graphql-ws subscription protocol from deprecated
+ yesod example with node client using graphql-ws npm package.
+ Bump version to v0.28.0

This PR will cease support for the older protocol due to line, which changes the name of an event from 'data' to 'next':
https://github.com/morpheusgraphql/morpheus-graphql/pull/805/files#diff-d7568c77ceef4d42fcddd7e99c510f1d7d53592431ab6ff0027f1667b6de68deR160

Deprecation notice:
https://www.npmjs.com/package/graphql-ws#disclaimer

Graphql-ws protocol:
https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md

Deprecated subscriptions-transport-ws protocol:
https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md

nalchevanidze commented:
thanks @ropwareJB. could you please remove Bumping of version to v0.28.0. since i use workflow which handles it automatically based on PR.

nalchevanidze commented on examples/yesod-pubsub/package.yaml:
```diff
@@ -0,0 +1,62 @@
+name:                yesod-gql-example
```
morpheus-graphql-examples-yesod-pubsub

ropwareJB replied:
Done!

nalchevanidze commented:
https://github.com/morpheusgraphql/morpheus-graphql/blob/9ab7ba48e6d286555ec504319fd8088afee68efc/config/stack.yaml#L4-L12

please add `yesod-pubsub` in the list. that way automatic config generation include it.

nalchevanidze commented on morpheus-graphql-subscriptions/src/Data/Morpheus/Subscriptions/Apollo.hs:
```diff
@@ -84,7 +87,10 @@ instance FromJSON a => FromJSON (ApolloSubscription a) where
-        ApolloSubscription <$> o .:? "id" <*> o .: "type" <*> o .:? "payload"
+        ApolloSubscription
+          <$> o .:? "id"
+          <*> o .: "type"
```
since we have now  ApolloResponseType, would be great if use it directly at  line 78  `apolloType :: ApolloResponseType` ;)

ropwareJB replied:
WIP with last commit - will resume tomorrow. Will need to update various client code as well, as the DS is used not just for responses, but requests as well.

ropwareJB replied:
Done!

nalchevanidze commented on examples/yesod-pubsub/stack.yaml:
```diff
@@ -0,0 +1,72 @@
+# This file was automatically generated by 'stack init'
```
we don't really need local stack file, don't we ?

ropwareJB replied:
Deleted

nalchevanidze commented:
@ropwareJB great work ;) i left some comments there.  please run `npm run setup 8.10.7` an the end to make sure the package and stack files have right configurations.

nalchevanidze commented:
@ropwareJB  by the way if you could also add test cases for new `graphql-ws` protocol would be brilliant, however but not requirement :)

ropwareJB commented:
+ Fixed Test cases
+ Added Ping/Pong GQL Message types
+ Added Ping/Pong test case
+ Added WSOutputEvent ADT to capture WS DS updates + message back

nalchevanidze commented on morpheus-graphql-client/src/Data/Morpheus/Client/Fetch/WebSockets.hs:
```diff
@@ -101,19 +102,19 @@ decodeMessage :: A.FromJSON a => ByteString -> GQLClientResult a
         apolloId = Just uid
       }
 
 endMessage :: Text -> ApolloSubscription ()
```
will this still support old protocols? since some of the m are working with "start" and "stop". may we can adjust it depending which protocol client will use?

ropwareJB replied:
As it is, unfortunately not. Renaming the types to suit the newer protocol breaks all the others.

I'm not sure how we would update the code to support multiple protocols - possibly some ReaderT instance but that is too complicated for my skills to integrate into the current codebase I think.

nalchevanidze replied:
would it be then okey. if i take over the pr?

nalchevanidze replied:
will crate followup PR for that
